File: lib/services/local_storage_service.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

import '../data/initial_shopping_items.dart';
import '../models/shopping_item.dart';
import '../models/shopping_session.dart';

class LocalStorageService {
  static const String _budgetLimitKey = 'budget_limit';
  static const String _itemsKey = 'shopping_items';
  static const String _initialItemsSeededKey = 'initial_items_seeded_v1';

  Future<ShoppingSession> loadSession() async {
    final preferences = await SharedPreferences.getInstance();
    final rawItems = await _loadOrSeedItems(preferences);

    return ShoppingSession(
      budgetLimit: preferences.getDouble(_budgetLimitKey),
      items: rawItems
          .map(
            (rawItem) => ShoppingItem.fromJson(
              jsonDecode(rawItem) as Map<String, dynamic>,
            ),
          )
          .toList(),
    );
  }

  Future<List<String>> _loadOrSeedItems(SharedPreferences preferences) async {
    final savedItems = preferences.getStringList(_itemsKey);
    final initialItemsSeeded =
        preferences.getBool(_initialItemsSeededKey) ?? false;

    if (initialItemsSeeded || (savedItems?.isNotEmpty ?? false)) {
      if (!initialItemsSeeded) {
        await preferences.setBool(_initialItemsSeededKey, true);
      }
      return savedItems ?? const <String>[];
    }

    final initialItems = InitialShoppingItems.create()
        .map((item) => jsonEncode(item.toJson()))
        .toList();

    await preferences.setStringList(_itemsKey, initialItems);
    await preferences.setBool(_initialItemsSeededKey, true);
    return initialItems;
  }

  Future<void> saveSession(ShoppingSession session) async {
    final preferences = await SharedPreferences.getInstance();

    await preferences.setStringList(
      _itemsKey,
      session.items.map((item) => jsonEncode(item.toJson())).toList(),
    );
    await preferences.setBool(_initialItemsSeededKey, true);

    if (session.budgetLimit == null) {
      await preferences.remove(_budgetLimitKey);
      return;
    }

    await preferences.setDouble(_budgetLimitKey, session.budgetLimit!);
  }
}

File: src/services/localStorageService.ts
import { ShoppingItem, ShoppingSession } from '../types';
import { InitialShoppingItems } from '../data/initialShoppingItems';

const BUDGET_LIMIT_KEY = 'budget_limit';
const ITEMS_KEY = 'shopping_items';
const INITIAL_ITEMS_SEEDED_KEY = 'initial_items_seeded_v1';

export class LocalStorageService {
  loadSession(): ShoppingSession {
    try {
      const budgetLimitRaw = localStorage.getItem(BUDGET_LIMIT_KEY);
      const budgetLimit = budgetLimitRaw ? parseFloat(budgetLimitRaw) : null;
      const rawItems = this.loadOrSeedItems();

      return {
        budgetLimit: isNaN(budgetLimit ?? NaN) ? null : budgetLimit,
        items: rawItems,
      };
    } catch {
      return {
        budgetLimit: null,
        items: InitialShoppingItems.create(),
      };
    }
  }

  private loadOrSeedItems(): ShoppingItem[] {
    const savedItemsRaw = localStorage.getItem(ITEMS_KEY);
    const initialItemsSeeded = localStorage.getItem(INITIAL_ITEMS_SEEDED_KEY) === 'true';

    if (initialItemsSeeded || (savedItemsRaw && savedItemsRaw.length > 0)) {
      if (!initialItemsSeeded) {
        localStorage.setItem(INITIAL_ITEMS_SEEDED_KEY, 'true');
      }
      try {
        const parsed = JSON.parse(savedItemsRaw || '[]');
        return Array.isArray(parsed) ? parsed : [];
      } catch {
        return [];
      }
    }

    const initial = InitialShoppingItems.create();
    localStorage.setItem(ITEMS_KEY, JSON.stringify(initial));
    localStorage.setItem(INITIAL_ITEMS_SEEDED_KEY, 'true');
    return initial;
  }

  saveSession(session: ShoppingSession): void {
    try {
      localStorage.setItem(ITEMS_KEY, JSON.stringify(session.items));
      localStorage.setItem(INITIAL_ITEMS_SEEDED_KEY, 'true');

      if (session.budgetLimit === null || session.budgetLimit === undefined) {
        localStorage.removeItem(BUDGET_LIMIT_KEY);
      } else {
        localStorage.setItem(BUDGET_LIMIT_KEY, session.budgetLimit.toString());
      }
    } catch (e) {
      console.error('Failed to save session to localStorage:', e);
    }
  }
}

export const localStorageService = new LocalStorageService();
